### DISCUSSION TITLE
Icons in taskbar non-immersive menus

### DISCUSSION THREAD

**Discussion by Anixx** [valinet/ExplorerPatcher#1113] [General]
Here is a comparison of icons in EP context menus versus Taskbar-context-menus-tweaker.

EP:
![изображение](https://user-images.githubusercontent.com/1271736/226127262-814797b2-48ab-4d41-b99e-5eab31b0f340.png)

Taskbar-context-menus-tweaker:
![изображение](https://user-images.githubusercontent.com/1271736/226126737-6771bd2c-9b46-42ae-bac6-a45d263075bd.png)

The Taskbar-context-menus-tweaker variant looks better to me.

- It has more items with icons.

- It does not have empty space to the left of icons (obviously, it is the place where the icons should be but invisible).

When the both utilities are on, the both sets of icons are visible, which proves that in EP variant the icons are simply not shown where they should be:

![изображение](https://user-images.githubusercontent.com/1271736/226129328-e52c2166-2767-4a44-b73d-a25495181e7e.png)

**Comment by valinet** [comment]
Don't we have a thread for the classic theme?

**Reply by Anixx** [reply]
This affects all themes, it appears when people use non-immersive menus.
![изображение](https://user-images.githubusercontent.com/1271736/226129606-0498454e-27e6-47df-8ab7-f4b433631b42.png)

**Reply by valinet** [reply]
Cannot reproduce.

![image](https://user-images.githubusercontent.com/6503598/226129739-afb6a574-1431-4a78-8743-56bfc72a3fca.png)

**Reply by Anixx** [reply]
The above screenshot is from Non-immersive-taskbar-context-menu mod for Windhawk. It is based on Taskbar-context-menus-tweaker. https://github.com/ramensoftware/windhawk-mods/blob/97c8aee81a71a356f145b60251c26353197255a4/mods/classic-taskbar-context-menu.wh.cpp 

My point is, it has more icons and better aligned. 

The utility has problems with second-level menus though (as opposed to EP).

**Reply by Anixx** [reply]
![изображение](https://user-images.githubusercontent.com/1271736/226130003-335ad04b-d582-4c72-ba9b-73cfe66af9a9.png)

**Reply by valinet** [reply]
Okay, and? ExplorerPatcher just removes the skin, as the option states. I do not add any icons, never stated that. Alignment is more than fine for me, probably (again) an issue with the "classic theme".

**Reply by Anixx** [reply]
When the both utilities are running, the icons appear where EP has empty space:
![изображение](https://user-images.githubusercontent.com/1271736/226130131-6819f411-d5e9-4652-a01b-94336b9d9485.png)

It looks like the icons are simply invisibe when EP is running (it fixes the text but not icons). Also, I wonder whether the gear icon is added by EP?

**Reply by valinet** [reply]
No, the gear icon is provided by Windows. EP only disables the custom drawing. If it works in Windhawk, why don't you disable the option in EP...?

**Reply by Anixx** [reply]
Because Windhawk has an issue with 2nd-level menus, just as Taskbar-context-menus-tweaker did before EP:

![изображение](https://user-images.githubusercontent.com/1271736/226130435-2a246964-e074-4f87-bc0c-07d16ddd5406.png)

**Reply by valinet** [reply]
Then contact the module author to fix it, or try to fix it yourself, I see the source code is published, what is the problem?

**Reply by Anixx** [reply]
I don't know programming. But could you at least remove that gear icon? It even does not fit, touching the text. After this, the simultanious use of the both utilities will produce the desired result.

**Reply by valinet** [reply]
> I don't know programming. But could you at least remove that gear icon? It even does not fit, touching the text. After this, the simultanious use of the both utilities will produce the desired result.

I don't know how and do not have time to do it, sorry.

**Reply by Anixx** [reply]
In other context menus there is no space on the left of the icons:
![изображение](https://user-images.githubusercontent.com/1271736/226143368-675c2e38-2f10-4ed5-8bfa-51daa1f4bd6d.png)

This looks correctly.

**Reply by Anixx** [reply]
Here is implementation of cogwheel removal feature from another, new Windhawk mod:
```
using SetMenuItemInfoW_t = decltype(&SetMenuItemInfoW);
SetMenuItemInfoW_t SetMenuItemInfoW_orig;
BOOL SetMenuItemInfoW_hook(
    HMENU            hMenu,
    UINT             item,
    BOOL             fByPosition,
    LPCMENUITEMINFOW lpmii
)
{
    if ((!fByPosition && item == 413) && settings.nosettingsicon) // LE HOMESTUCK??
    {
        return TRUE;
    }

    return SetMenuItemInfoW_orig(
        hMenu,
        item,
        fByPosition,
        lpmii
    );
}
```

**Reply by pyrates999** [reply]
If windhawk is able to do this, you should install that and enable that mod.
